cluster-autoscaler/cloudprovider/aws/auto_scaling_groups.go

```diff
@@ -18,80 +18,168 @@ package aws
 	"github.com/golang/glog"
 )
 
+const scaleToZeroSupported = false
+
 type autoScalingGroups struct {
-	registeredAsgs           []*asgInformation
-	instanceToAsg            map[AwsRef]*Asg
-	cacheMutex               sync.Mutex
-	instancesNotInManagedAsg map[AwsRef]struct{}
-	service                  autoScalingWrapper
+	registeredAsgs      []*asgInformation
+	instanceToAsg       map[AwsRef]*Asg
+	notInRegisteredAsg  map[AwsRef]bool
+	registeredAsgsMutex sync.RWMutex
+	instanceToAsgMutex  sync.Mutex
+	service             autoScalingWrapper
+	interrupt           chan struct{}
 }
 
-func newAutoScalingGroups(service autoScalingWrapper) *autoScalingGroups {
+func newAutoScalingGroups(service autoScalingWrapper) (*autoScalingGroups, error) {
 	registry := &autoScalingGroups{
-		registeredAsgs:           make([]*asgInformation, 0),
-		service:                  service,
-		instanceToAsg:            make(map[AwsRef]*Asg),
-		instancesNotInManagedAsg: make(map[AwsRef]struct{}),
+		registeredAsgs:     make([]*asgInformation, 0),
+		service:            service,
+		instanceToAsg:      make(map[AwsRef]*Asg),
+		notInRegisteredAsg: make(map[AwsRef]bool),
+		interrupt:          make(chan struct{}),
 	}
-	return registry
+	go wait.Until(func() {
+		if err := registry.regenerateCache(); err != nil {
+			glog.Errorf("Error while regenerating Asg cache: %v", err)
+		}
+	}, time.Hour, registry.interrupt)
+
+	return registry, nil
 }
 
-// Register registers asg in Aws Manager.
-func (m *autoScalingGroups) Register(asg *Asg) {
-	m.cacheMutex.Lock()
-	defer m.cacheMutex.Unlock()
+// Register ASG. Returns true if the ASG was registered.
+func (m *autoScalingGroups) Register(asg *Asg) bool {
+	m.registeredAsgsMutex.Lock()
+	defer m.registeredAsgsMutex.Unlock()
 
+	for i := range m.registeredAsgs {
+		if existing := m.registeredAsgs[i].config; existing.AwsRef == asg.AwsRef {
+			if reflect.DeepEqual(existing, asg) {
+				return false
+			}
+			m.registeredAsgs[i].config = asg
+			glog.V(4).Infof("Updated ASG %s", asg.AwsRef.Name)
+			m.invalidateUnownedInstanceCache()
+			return true
+		}
+	}
+
+	glog.V(1).Infof("Registering ASG %s", asg.AwsRef.Name)
 	m.registeredAsgs = append(m.registeredAsgs, &asgInformation{
 		config: asg,
 	})
+	m.invalidateUnownedInstanceCache()
+	return true
+}
+
+// Unregister ASG. Returns true if the ASG was unregistered.
+func (m *autoScalingGroups) Unregister(asg *Asg) bool {
+	m.registeredAsgsMutex.Lock()
+	defer m.registeredAsgsMutex.Unlock()
+
+	updated := make([]*asgInformation, 0, len(m.registeredAsgs))
+	changed := false
+	for _, existing := range m.registeredAsgs {
+		if existing.config.AwsRef == asg.AwsRef {
+			glog.V(1).Infof("Unregistered ASG %s", asg.AwsRef.Name)
+			continue
+		}
+		updated = append(updated, existing)
+		changed = true
+	}
+	m.registeredAsgs = updated
+	return changed
+}
+
+func (m *autoScalingGroups) get() []*asgInformation {
+	m.registeredAsgsMutex.RLock()
+	defer m.registeredAsgsMutex.RUnlock()
```

Comment:
You're correct that we don't mutate `registeredAsgs` here, but we do elsewhere (i.e. in `Register` and `Unregister`. We need to take a lock when reading memory - not only when writing it. This `RLock` is intended to prevent us reading `registeredAsgs` while another goroutine is writing to it.

For example if `play.go` were as follows:
```go
package main

import (
	"fmt"
	"sync"
	"time"
)

func main() {
	s := make([]int, 0)
	m := sync.RWMutex{}
	go func() {
		for i := 0; ; i++ {
			m.Lock()
			s = append(s, i)
			m.Unlock()
			time.Sleep(1 * time.Second)
		}
	}()
	for {
		fmt.Println(s)
		time.Sleep(1 * time.Second)
	}
}
```

```bash
$ go run -race play.go
[]           
==================                                
WARNING: DATA RACE
Write at 0x00c420092020 by goroutine 6:
  main.main.func1()
      /Users/negz/control/go/src/play.go:15 +0xf4
         
Previous read at 0x00c420092020 by main goroutine:
  main.main()
      /Users/negz/control/go/src/play.go:21 +0x158
                 
Goroutine 6 (running) created at:
  main.main()        
      /Users/negz/control/go/src/play.go:12 +0x147
==================  
```

Adding an `m.RLock()` and `m.RUnlock()` around `fmt.Println(s)` prevents the data race.

Note that maybe I'm missing something and these methods are in fact only ever called by a single goroutine, in which case all this locking is pointless. :)

Reply:
It seems like I had not understood the use-case of RLock/RUnlock.
Thanks for the clear explanation!